### PR TITLE
Allow to dynamically add strategy in the plugin init method

### PR DESCRIPTION
**Pull request by Aschen**
## What does this PR do ?

When dynamically adding strategies with the `accessors.strategies.add` method in the plugin `init` method, the authenticators have not been initialized yet.  

This PR initialize statically declared resources (controllers, authenticators, pipes, etc.) before calling the plugin `init` method.

https://deploy-preview-1479--doc-kuzzle.netlify.com/plugins/guides/strategies/overview/

### How should this be manually tested?

Test plugin:
```
class TestPlugin {
  constructor () {
    this.authenticators = {
      'ExampleStrategy': require('passport-custom').Strategy
    };
  }

  init (config, context) {
    return context.accessors.strategies.add('test',  {
      config: {
        authenticator: 'ExampleStrategy'
      },
      methods: {
        create: 'create',
        delete: 'delete',
        exists: 'exists',
        getById: 'getById',
        getInfo: 'getInfo',
        update: 'update',
        validate: 'validate',
        verify: 'verify'
      }
    })
  }

  create () {}
  exists () {}
  delete () {}
  getById () {}
  getInfo () {}
  update () {}
  validate () {}
  verify () {}
}

module.exports = TestPlugin
```

### CODE REVIEW

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L972](https://github.com/kuzzleio/kuzzle/blob/2d234e681fab5f0af2f5752a33dccc3b00506ccb/lib/api/core/plugins/pluginsManager.js#L972): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L972](https://github.com/kuzzleio/kuzzle/blob/148afc6db4a87ea85dd3cadae3b3b9c0fcd02fdf/lib/api/core/plugins/pluginsManager.js#L972): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [pluginsManager.js#L999](https://github.com/kuzzleio/kuzzle/blob/bbc322417c2e265c39e65db1a4f38c1daeeed82a/lib/api/core/plugins/pluginsManager.js#L999): Group all shorthand properties at either the beginning or end of this object declaration. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3499)
1. ![INFO][INFO] [pluginsManager.js#L920](https://github.com/kuzzleio/kuzzle/blob/bbc322417c2e265c39e65db1a4f38c1daeeed82a/lib/api/core/plugins/pluginsManager.js#L920): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Aschen** [comment]
@scottinet This is the solution I'm working on

File: lib/api/core/plugins/pluginsManager.js [hunk 1]
```diff
@@ -569,6 +574,16 @@ class PluginsManager {
    *                                     registration fails
    */
   registerStrategy(pluginName, strategyName, strategy) {
+    // only add the strategy to the strategies object if the init method
+    // has not been called
+    if (! this.plugins[pluginName].initCalled) {
+      this.plugins[pluginName].object.strategies =
+        this.plugins[pluginName].object.strategies || {};
+        this.plugins[pluginName].object.strategies[strategyName] = strategy;
```
Comment (Aschen) [inline lib/api/core/plugins/pluginsManager.js]:
```suggestion
      this.plugins[pluginName].object.strategies[strategyName] = strategy;
```

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [pluginsManager.js#L999](https://github.com/kuzzleio/kuzzle/blob/c03b068cbb3f720df3fa99b97786a054932c9d9c/lib/api/core/plugins/pluginsManager.js#L999): Group all shorthand properties at either the beginning or end of this object declaration. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3499)
1. ![INFO][INFO] [pluginsManager.js#L920](https://github.com/kuzzleio/kuzzle/blob/c03b068cbb3f720df3fa99b97786a054932c9d9c/lib/api/core/plugins/pluginsManager.js#L920): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [pluginsManager.js#L998](https://github.com/kuzzleio/kuzzle/blob/a2e44955b0ca183eebb078dcd2149dea0927ccd6/lib/api/core/plugins/pluginsManager.js#L998): Group all shorthand properties at either the beginning or end of this object declaration. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3499)
1. ![INFO][INFO] [pluginsManager.js#L919](https://github.com/kuzzleio/kuzzle/blob/a2e44955b0ca183eebb078dcd2149dea0927ccd6/lib/api/core/plugins/pluginsManager.js#L919): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [pluginsManager.js#L997](https://github.com/kuzzleio/kuzzle/blob/c9b55daac00efbbcdea956d073967d81e8c12fbb/lib/api/core/plugins/pluginsManager.js#L997): Group all shorthand properties at either the beginning or end of this object declaration. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3499)
1. ![INFO][INFO] [pluginsManager.js#L918](https://github.com/kuzzleio/kuzzle/blob/c9b55daac00efbbcdea956d073967d81e8c12fbb/lib/api/core/plugins/pluginsManager.js#L918): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Aschen** [comment]
@scottinet I added some documentation, you might want to review this PR again

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [pluginsManager.js#L997](https://github.com/kuzzleio/kuzzle/blob/e304d531ceae9d1736dad8c17c208855db1dbe77/lib/api/core/plugins/pluginsManager.js#L997): Group all shorthand properties at either the beginning or end of this object declaration. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3499)
1. ![INFO][INFO] [pluginsManager.js#L918](https://github.com/kuzzleio/kuzzle/blob/e304d531ceae9d1736dad8c17c208855db1dbe77/lib/api/core/plugins/pluginsManager.js#L918): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: lib/api/core/plugins/pluginsManager.js [hunk 2]
```diff
@@ -982,6 +995,7 @@ class PluginsManager {
 
       const
         plugin = {
```
Comment (kuzzle) [inline lib/api/core/plugins/pluginsManager.js]: MINOR Group all shorthand properties at either the beginning or end of this object declaration. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3499)

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L918](https://github.com/kuzzleio/kuzzle/blob/2329324830f552d23bdffd2b06c519dfd55c9e01/lib/api/core/plugins/pluginsManager.js#L918): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L918](https://github.com/kuzzleio/kuzzle/blob/9ed1dda2f03329372926ad7462fff08ede85dc2f/lib/api/core/plugins/pluginsManager.js#L918): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L918](https://github.com/kuzzleio/kuzzle/blob/bbe039d153b544ccdaeb090def9a90eb3d25fd4d/lib/api/core/plugins/pluginsManager.js#L918): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: doc/1/plugins/guides/strategies/overview/index.md [hunk 3]
```diff
@@ -28,42 +28,81 @@ this.authenticators = {
...
+this.authenticators = {
+  Local: require('passport-local')
+};
+
+this.strategies = {
+  local: {
+    config: {
+      authenticator: 'Local'
+    },
+    // theses methods must exists in the plugin object
```
Comment (scottinet) [inline doc/1/plugins/guides/strategies/overview/index.md]:
```suggestion
    // these methods must be exposed by the plugin 
```

File: doc/1/plugins/guides/strategies/overview/index.md [hunk 4]
```diff
@@ -28,42 +28,81 @@ this.authenticators = {
...
+Whether strategies are added statically or dynamically, the registered strategy must expose the following properties:
 
 | Arguments | Type              | Description                           |
 | --------- | ----------------- | ------------------------------------- |
 | `config`  | <pre>object</pre> | Authentication strategy configuration |
 | `methods` | <pre>object</pre> | List of exposed methods               |
 
+### statically register strategies
+
+Plugins can declare a `strategies` object which contain authentication strategies to register.
```
Comment (scottinet) [inline doc/1/plugins/guides/strategies/overview/index.md]: typo + suggestion:

```suggestion
Plugins can declare a `strategies` object which contains the authentication strategies to register.  
This object will be interpreted by Kuzzle only once, immediately after this plugin's init function has resolved.  
```

File: doc/1/plugins/guides/strategies/overview/index.md [hunk 5]
```diff
@@ -28,42 +28,81 @@ this.authenticators = {
...
+  }
+}
+```
+
+### dynamically register strategies
+
+Strategies can be register at runtime with the [strategies.add](/core/1/plugins/plugin-context/accessors/strategies/#add) method.  
+
+::: info
+Strategies added dynamically in the plugin [init method](/core/1/plugins/guides/manual-setup/init-function/) are added to the static `strategies` object and loaded by Kuzzle after plugin initialization.
```
Comment (scottinet) [inline doc/1/plugins/guides/strategies/overview/index.md]:
```suggestion
Strategies added dynamically in the plugin's [init method](/core/1/plugins/guides/manual-setup/init-function/) are added to the static `strategies` object and loaded by Kuzzle after the plugin initialization.
```

**Comment by kuzzle** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [pluginsManager.js#L918](https://github.com/kuzzleio/kuzzle/blob/2f948bf0b6fe5da8fc2bed56a48549ff8bd4ed39/lib/api/core/plugins/pluginsManager.js#L918): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Pull request merged**